[FEATURE] Introduction of QGIS Quick library

PeterPetrik opened the pull request:
[FEATURE] Introduction of QGIS Quick library

This pull request is a subset of https://github.com/qgis/QGIS/pull/6490

This adds a new library for creation of applications based on Qt Quick
framework.
It contains reusable QML / Qt Quick components based on QGIS core
library.
The initial work introduces MapCanvas

To enable compilation of the library, use WITH_QUICK=TRUE

Further documentation of the library is located in doc/qgsquick.dox

For background information see the associated QEP:
qgis/QGIS-Enhancement-Proposals#109

The initial implementation is largely based on the work of Matthias Kuhn
and Marco Bernasocchi on QField probject - kudos to them for the great
job!

jdugge committed:
[FEATURE] Make map tips delay configurable

jdugge committed:
#5416 Add map tips delay setting to menu

jdugge committed:
#5416 Update maptips delay on saving the settings

jdugge committed:
#5416 Remove spurious settings

jdugge committed:
#5416 Re-insert update of map tips delay

Commit:
Fix geroreferencer crash when deleting a point through context menu

PeterPetrik committed:
[FEATURE] QgsMeshLayer part 1: Reading raw mesh

Introducting MDAL, QgsMeshLayer, mesh data providers (mesh_memory, mdal)
to read and visualize raw meshes: vertices and faces. Support dragging
2dm files from browser on canvas to visualize 2dm meshes.
Support for QgsMeshLayer in Python API.

DelazJ committed:
Replace Edit Nodes Item icon

nyalldawson committed:
Add geometry method to calculate mid point on an arc from p1->p2 with given center

And improve some docstrings

nyalldawson committed:
Add convenience constructor to create QgsLineString between two points

nyalldawson committed:
Add convenience constructors to create QgsCircularStrings

Add constructors to create a circular string with a single arc:
- from 3 points on arc
- from 2 points and a center

nyalldawson committed:
Allow using the longer arc with two point and center methods

nyalldawson committed:
Add QgsGeometry method to create wedge shaped buffers

Creates a wedge shaped buffer using circular strings, with
parameters for azimuth, wedge width (in degrees), outer radius
and inner radius.

nyalldawson committed:
[FEATURE][processing] Native wedge buffer algorithm

This algorithm creates wedge shaped buffers from input
points.

The azimuth parameter gives the angle (in degrees) for the
middle of the wedge to point. The buffer width (in degrees)
is specified by the width parameter. Note that the
wedge will extend to half of the angular width either side
of the azimuth direction.

The outer radius of the buffer is specified via outer radius,
and optionally an inner radius can also be specified.

Supports dynamic parameters for azimuth, width, and radius.

The native output from this algorithm is CurvePolygon
geometries, but these may be automatically segmentized to
Polygons depending on the output format.

nyalldawson committed:
Add unit test for wedge buffers alg

nyalldawson committed:
[FEATURE] Add wedge_buffer expression function

Returns a wedge shaped buffer originating from a point geometry,
with arguments for azimuth, buffer width (in degrees), outer radius
and inner radius.

nyalldawson committed:
Add some useful QgsVector methods

nyalldawson committed:
Add QgsLineSegment2D class, for simple 2d line segments which
consist of just a 2D start and end point

nyalldawson committed:
Declare typeinfo for QgsVector to optimise use in Qt containers

nyalldawson committed:
Apply clazy optimisation to orthogonalize algorithm

nyalldawson committed:
Add left of line test to QgsLineSegment2D

nyalldawson committed:
Add reverse method to QgsLineSegment2D

nyalldawson committed:
Add QgsRay2D class representing an infinite ray from a point in a direction

nyalldawson committed:
Add class for comparing distance from two line segments to an origin

nyalldawson committed:
Add class for comparing angles to points in a clockwise fashion

nyalldawson committed:
Add some more line segment methods

nyalldawson committed:
[oracle] Add read-only support for curved geometry types

Previously these geometries were silently discarded

Fixes #12675, #16809

Sponsored by Open Spatial (http://www.openspatial.com)

PeterPetrik committed:
fix build errors, use MDAL 0.0.2 (int API)

PeterPetrik committed:
use smart pointer internally and remove unnecessary code

PeterPetrik committed:
use QStringLiteral

PeterPetrik committed:
fix Doxygen warnings

PeterPetrik committed:
use unique_ptrs for tmp pointers

nyalldawson committed:
[oracle] Fixes for edit tests (still disabled)

nyalldawson committed:
[oracle] Always check bounding box intersection when using filter rect

Because oracle sdo_filter sometimes returns results outside
the given filter (depending on the spatial index), and is
only designed for use as a fast "first-pass" filter.

From the Oracle docs:

"This operator performs only a primary filter operation..

The secondary filtering operation, performed by the SDO_RELATE operator,
can be used to determine with certainty if objects interact spatially."

Instead of using SDO_RELATE we can instead just do a bounding
box intersection inside QGIS if an exact intersection is not
required.

Fixes failing provider conformance test suite for oracle provider.

nirvn committed:
[FEATURE][ogr/gdal] Add UI to expose HTTP(S) and cloud datasets

PeterPetrik committed:
[FEATURE] Introduction of QGIS Quick library

This adds a new library for creation of applications based on Qt Quick
framework. It contains reusable QML / Qt Quick components based on QGIS core
library. The initial work introduces all infrastrucure and MapCanvas and
MapSettings components.

To enable compilation of the library, use WITH_QUICK=TRUE

Further documentation of the library is located in doc/qgsquick.dox

For background information see the associated QEP:
qgis/QGIS-Enhancement-Proposals#109

The initial implementation is largely based on the work of Matthias Kuhn
and Marco Bernasocchi on QField probjectjob!

m-kuhn commented on src/core/qgsapplication.h:
```diff
@@ -405,6 +405,12 @@ class CORE_EXPORT QgsApplication : public QApplication
     static QString libexecPath();
 
+    /**
+     * Returns the path where QML components are installed for QGIS Quick library
```
Can you document what is returned if it's built without quick support?

PeterPetrik replied:
It returns CMake's (DEFAULT_)QML_SUBDIR regardless of if there is something in the directory (==WITH_QUICK) or not. Would it make sense to add here something like: "Could be empty" or so?

PeterPetrik replied:
I was referring to this https://github.com/PeterPetrik/QGIS/blob/qgsquick_1_canvas/CMakeLists.txt#L668

PeterPetrik replied:
fixed, returns empty QString and docs improved

m-kuhn commented on src/quickgui/plugin/CMakeLists.txt:
```diff
@@ -0,0 +1,121 @@
+)
+TARGET_LINK_LIBRARIES(qgis_quick_plugin qgis_quick)
+SET_TARGET_PROPERTIES(qgis_quick_plugin PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${QGIS_QUICK_PLUGIN_RUNTIME_DIR})
+TARGET_COMPILE_DEFINITIONS(qgis_quick_plugin PRIVATE "-DQUICK_EXPORT=" "-DDQT_NO_FOREACH")
```
Is the second `D` intended `-DDQT_NO_FOREACH`?

PeterPetrik replied:
no, typo. good catch. fixed

m-kuhn commented on src/quickgui/plugin/qgsquickplugin.h:
```diff
@@ -0,0 +1,43 @@
+class QgisQuickPlugin : public QQmlExtensionPlugin
+{
+    Q_OBJECT
+    Q_PLUGIN_METADATA( IID "org.qt-project.Qt.QQmlExtensionInterface" )
```
Should this rather refer to QGIS than to the Qt Project?

PeterPetrik replied:
The `Q_PLUGIN_METADATA` references a Qt interface required for Qt Plugin system. This is  [QmlExtension](http://doc.qt.io/qt-5/qqmlextensionplugin.html) plugin type, so we define here that QgisQuickPlugin is a QML Qt plugin.

m-kuhn commented on src/quickgui/qgsquickmapcanvasmap.h:
```diff
@@ -0,0 +1,172 @@
+    //! freeze changed
+    void freezeChanged();
+
+    //! is rendering changed
```
I am not thrilled by these doxygen comments. I know it's not fun to write those for methods and signals, one didn't write himself.

Note, what I normally do is to really think about the *property* and forget about the setter/getter/signal part. That is I'd copy paste for getter/signal of `isRendering` (and setter for others where applicable) a doxygen string like

```
/**
 * The isRendering property is set to true while a rendering job is pending for this map canvas map.
 * It can be used to show a notification icon about an ongoing rendering job.
 * This is a readonly property.
 */
```

PeterPetrik replied:
I am sorry, but writing documentation is not something I feel comfortable with. I tried to improve the docs based on your suggestion where possible. Copy of documentation was done by [\copydoc](https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdcopydoc) where possible

m-kuhn commented on src/quickgui/qgsquickmapsettings.cpp:
```diff
@@ -0,0 +1,224 @@
+  {
+    connect( mProject, &QgsProject::readProject, this, &QgsQuickMapSettings::onReadProject );
+
+    // TODO: we have a problem here, since project can have alread .qgs loaded, so
```
already

PeterPetrik replied:
already

m-kuhn commented on src/quickgui/qgsquickmapsettings.cpp:
```diff
@@ -0,0 +1,224 @@
+    // TODO: we have a problem here, since project can have alread .qgs loaded, so
+    // we are unable to get DOM to reload project settings for the canvas
+    // fortunately setProject is used only once at the very beginning of the run,
+    // so it should work ....
```
Was that me who wrote this comment? I'm having a hard time understanding what I wanted to tell the world :D

PeterPetrik replied:
it was someone over here, probably me. The problem arose with removal of `QgsProject::instance`.

Bug is when you  do something like 
```
QgsProject p;
p.read()
qgsquickmapsettings.setProject(p);
```
and in this case QgsQuickMapSettings::onReadProject is not triggered, so it is not possible to read mapcanvas settings from DOM

m-kuhn replied:
And if one would just trigger an onReadProject() here?

PeterPetrik replied:
you do not have DOM string any more.

m-kuhn replied:
It's been a while since I wrote all this... Why are things read directly from the DOM? Are they not available elsewhere?

PeterPetrik replied:
But the parameter is QgsProject, and you do not have any other mapsettings instance but internal one in the qgsquickmapsettings... not sure from where we could read it?

m-kuhn replied:
That's what I'm asking you.
If it's not possible (could be) I think the TODO comment can be removed (and possibly replaced with a comment for pre-conditions).

PeterPetrik replied:
It is not possible without some refactoring of QgsProject. Moved comment to header for now to state prerequisite.

m-kuhn commented on tests/src/quickgui/app/main.cpp:
```diff
@@ -0,0 +1,97 @@
+  // Add some data for debugging if needed
+  QgsApplication::messageLog()->logMessage( QgsQuickUtils::instance()->dumpScreenInfo() );
+  QgsApplication::messageLog()->logMessage( "data directory: " + dataDir );
+  QgsApplication::messageLog()->logMessage( "All up and running" );
```
I think what ends up in the message log should be translated.

PeterPetrik replied:
does it apply to tests too?

m-kuhn replied:
I don't think so, but there you'd probably use QgsDebugMsg (no reason to send it through the message log).

PeterPetrik replied:
Since this is kind-of test (in test/ directory), I do not know how it works but I would not like to add translators more work for some random string in test cases.

m-kuhn replied:
ah, you are right. Forget it.
Can you have a look at the dump method in the QgsQuickUtils then?

PeterPetrik replied:
I am sorry, which method do you mean? If dumpScreenInfo, what is wrong with it?

m-kuhn replied:
I think it doesn't live in test and I think it is untranslated, no?

PeterPetrik replied:
translated

m-kuhn commented on tests/src/quickgui/app/main.cpp:
```diff
@@ -0,0 +1,97 @@
+
+  // Set simulated position for desktop builds
+  bool use_simulated_position = true;
+  engine.rootContext()->setContextProperty( "__use_simulated_position", use_simulated_position );
```
Can we skip that until positioning is discussed in a followup pull request?

PeterPetrik replied:
removed

m-kuhn commented on tests/src/quickgui/app/main.cpp:
```diff
@@ -0,0 +1,97 @@
+  QQmlEngine engine;
+  engine.addImportPath( QgsApplication::qmlImportPath() );
+  engine.rootContext()->setContextProperty( "__project", &project );
+  engine.rootContext()->setContextProperty( "__layers", QVariant::fromValue( project.layerTreeRoot()->layerOrder() ) );
```
Is that something that can be accessed through the project?
I'm afraid of having something highly dynamic like the layer list as a context property.

PeterPetrik replied:
`layerTreeRoot()`, nor `layerOrder()` is INVOCABLE. So I would need to change both of those classes plus register another types in qml plugin, or create a function in qgsquickutils.cpp. Since this is just a test (in real project one would need either project chooser or some layer tree item), I am not sure if we want to make such changes in core libraries?

m-kuhn replied:
ah right, it's a test again, forget it

PeterPetrik replied:
forgotten

PeterPetrik commented on src/quickgui/qgsquickutils.cpp:
```diff
@@ -0,0 +1,77 @@
+  if ( !sInstance )
+  {
+    QgsDebugMsg( QStringLiteral( "QgsQuickUtils created: %1" ).arg( long( QThread::currentThreadId() ) ) );
+    sInstance = new QgsQuickUtils();
```
I do not think so: https://github.com/qgis/QGIS/pull/6842/files/ca3243d8656e0431b2a0e00803db717e580d22d6#diff-09ffc66fca3d9120f66c33a81f485949R41

m-kuhn replied:
Is this the standard way to deal with singletons in QML context? I'd expect the `new` to happen where it is registered. With this approach here responsibilities for creation and cleanup are distributed in a non-intuitive way.

PeterPetrik replied:
Actually, we needed to set dp from C++ code at some point, but later we decided that it can be done automatically. So singleton in no longer needed in C++ context, so I implemented [QML singleton](http://doc.qt.io/qt-5/qqmlengine.html#qmlRegisterSingletonType) as suggested in QT docs.

DelazJ committed:
Fix labels in Displacement Point renderer GUI

nyalldawson committed:
Indentation

DelazJ committed:
Use "Displacement lines" label

NathanW2 committed:
[FEATURE][needs-docs] Rework expression parser for unknown functions (#6840)

* [FEATURE][needs-docs] Rework expression parser for unknown functions

* Remove old comment

* Rename NAME and update tests

* Change COLUMN_REF to QUOTED_COLUMN_REF

gacarrillor committed:
[processing] Make line intersection alg able to handle intersections of type 'geometry collection'

PeterPetrik committed:
qgisquick code cleaning

rldhont committed:
[Tests][Server] Add unit test for empty contextual legend

Manually port c217cbb4dd140695af74debbe052dbefc1d21e74

PeterPetrik committed:
add docs and move singleton to QML

PeterPetrik committed:
remove DQT_NO_FOREACH from app

(some core headers still use it)

PeterPetrik commented:
@m-kuhn thank you for your review, I addressed the comments.

daniviga committed:
Remove groups definition from RPM spec
since it has been deprecated (ref: https://fedoraproject.org/wiki/RPMGroups)

wonder-sk committed:
Port Difference & Sym.Diff. to C++, cleanups and tests

jef-n committed:
[postgres][db2] still save username when instructed too also with authcfg

(cherry picked from commit 4ad35311037f226880ac3dd30eec0f92d119ca1c)

jef-n committed:
fix astyle version check

(cherry picked from commit dee284fe974560a3c9723e35123a02c7c9d30b8a)

PeterPetrik committed:
fix warnings about Qt connection with nullptr sender

gacarrillor committed:
[processing] Adjust code for line intersection with collection intersections

daniviga committed:
Make RPM package compatible with Fedora 28
which ships GRASS 7.4

PeterPetrik committed:
another try to fix qt qml plugin dump

PeterPetrik commented:
AH finally I manage to fix that qmlplugindump executable and fix the build! @m-kuhn I hope now it is ready for re-review. Thanks

wonder-sk committed:
difference: skip empty geometries, convert to multi-part geometries

m-kuhn committed:
No print() usage in processing core

nyalldawson committed:
[oracle] Fix handling of NULL values when add features to provider

Fixes NULL attribute values are incorrectly converted to empty
strings or 0 numeric values instead of NULL.

nyalldawson committed:
[oracle] Fix incorrect feature count after adding features to layers

nyalldawson committed:
[oracle] Some updates to Oracle provider test suite

Now the Oracle provider passes the data provider edit stress/
conformance tests without issue!

nyalldawson committed:
Remove geos include from qgsgeometry.h

Avoids the need to include geos from some build targets

nyalldawson committed:
Remove some redundant geos includes

nyalldawson committed:
Remove some more redundant include directories

nyalldawson committed:
[mdal] Fix bad header guard

nyalldawson committed:
Fix doxygen warnings

nyalldawson committed:
Followups

PeterPetrik commented on tests/src/quickgui/app/CMakeLists.txt:
```diff
@@ -63,8 +63,6 @@ ADD_EXECUTABLE(${QGIS_QUICK_APP_NAME}
 TARGET_LINK_LIBRARIES(${QGIS_QUICK_APP_NAME} Qt5::Gui Qt5::Qml Qt5::Quick Qt5::Xml qgis_core qgis_quick)
 SET_TARGET_PROPERTIES(${QGIS_QUICK_APP_NAME} PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY})
-
-TARGET_COMPILE_DEFINITIONS(${QGIS_QUICK_APP_NAME} PRIVATE "-DDQT_NO_FOREACH")
```
agree completely, but I there is a FOREACH used in one of the qgis core library headers that are included in this. Do not remember in which one. But to be able to use it here, we would need to fix all qgis core headers first.

m-kuhn replied:
https://github.com/qgis/QGIS/pull/6862

m-kuhn replied:
Merged, let's see if we can enable that!

PeterPetrik replied:
let me try it

m-kuhn commented on src/core/qgsrelationmanager.cpp:
```diff
@@ -25,9 +25,12 @@ QgsRelationManager::QgsRelationManager( QgsProject *project )
-  connect( project, &QgsProject::layersRemoved, this, &QgsRelationManager::layersRemoved );
+  if ( mProject )
+  {
+    connect( project, &QgsProject::readProject, this, &QgsRelationManager::readProject );
```
Isn't QgsRelationManager always instantiated by QgsProject (and then owned by and a child of this project)?
It seems weird if the project can be a `nullptr` in this scenario. Did I miss something here?

PeterPetrik replied:
normally yes, but [qmlplugindump](http://doc.qt.io/qtcreator/creator-qml-modules-with-plugins.html#generating-qmltypes-files) uses default constructor to initialize types to be able to extract some info from it.

m-kuhn replied:
Out of curiosity, when is this executed, build time or run time?

PeterPetrik replied:
it is executed after target qgis_quick_plugin is build, see 
https://github.com/PeterPetrik/QGIS/blob/qgsquick_1_canvas/src/quickgui/plugin/CMakeLists.txt#L106

m-kuhn replied:
And if it's registered as uncreatable type (as done originally in QField)?

PeterPetrik replied:
yep, lets register just uncreatable type.

m-kuhn committed:
Rewrite all core header usages of Q_FOREACH

m-kuhn committed:
Also remove Q_FOREACH from all gui headers

m-kuhn committed:
Fix doxygen warning

PeterPetrik committed:
remove FOREACH

PeterPetrik commented:
hmm, did not meant that
